File: 203.Remove Linked List Elements.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
 /*
 An excellent solution shared at https://leetcode.com/discuss/47642/simple-and-elegant-solution-in-c
 ListNode *removeElements(ListNode *head, int val)
{
    ListNode **list = &head;

    while (*list != nullptr)
    {
        if ((*list)->val == val)
        {
            *list = (*list)->next;
        }
        else
        {
            list = &(*list)->next;
        }
    }

    return head;
}
 */
struct ListNode* removeElements(struct ListNode* head, int val) {
    struct ListNode* curr = head;
    struct ListNode* prev = NULL;
    while(curr!=NULL)
    {
        if(curr->val == val)
        {
            if(prev==NULL)
            {
                curr = curr->next;
                head = curr;
            }
            else
            {
                prev->next = curr->next;
                curr = curr->next;
            }
            continue;
        }
        prev = curr;
        curr = prev->next;
    }
    return head;
}

File: 203.Remove Linked List Elements.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} val
# @return {ListNode}
def remove_elements(head, val)
    dummy = ListNode.new(-1, head)
    slow = dummy
    fast = head
    while fast
        fast = fast.next while fast && fast.val == val
        if fast
            slow.next.val = fast.val
            fast = fast.next
            slow = slow.next
        end
    end
    slow.next = nil
    return dummy.next
end
